crates/sui-framework/packages/sui-system/tests/rewards_distribution_tests.move

```diff
@@ -142,32 +153,32 @@ module sui_system::rewards_distribution_tests {
         stake_with(STAKER_ADDR_1, VALIDATOR_ADDR_1, 100, scenario);
         stake_with(STAKER_ADDR_2, VALIDATOR_ADDR_2, 100, scenario);
         governance_test_utils::advance_epoch(scenario);
+        // V1: 200, V2: 300, V3: 300, V4: 400
 
         set_commission_rate_and_advance_epoch(VALIDATOR_ADDR_2, 5000, scenario); // 50% commission
-
-        // 10 SUI for each 100 SUI staked
         advance_epoch_with_reward_amounts(0, 120, scenario);
+        // V1: 220, V2: 330, V3: 330, V4: 440
         // 5 SUI, or 50 % of staker_2's rewards, goes to validator_2
-        assert_validator_non_self_stake_amounts(validator_addrs(), vector[110, 105, 0, 0], scenario);
-        assert_validator_self_stake_amounts(validator_addrs(), vector[110, 225, 330, 440], scenario);
+        assert_validator_non_self_stake_amounts(validator_addrs(), vector[110 * MIST_PER_SUI, 105 * MIST_PER_SUI, 0, 0], scenario);
+        assert_validator_self_stake_amounts(validator_addrs(), vector[110 * MIST_PER_SUI, 225 * MIST_PER_SUI, 330 * MIST_PER_SUI, 440 * MIST_PER_SUI], scenario);
 
         set_commission_rate_and_advance_epoch(VALIDATOR_ADDR_1, 1000, scenario); // 10% commission
 
-        // 20 SUI for each 110 SUI staked
-        advance_epoch_with_reward_amounts(0, 240, scenario);
-
-        // 2 SUI, or 10 % of staker_1's rewards (20 SUI), goes to validator_1
-        // so staker_1 now has 110 + 20 - 2 = 128 SUI.
-        // And 10 SUI, or 50% of staker_2's rewards (20 SUI) goes to validator_2
-        // so staker_2 now has 105 +20 - 10 = 115 SUI.
-        assert_validator_non_self_stake_amounts(validator_addrs(), vector[128, 115, 0, 0], scenario);
-
-        // validator_1 gets 20 SUI of their own rewards and 2 SUI of commission
-        // so in total 110 + 20 + 2 = 132 SUI.
-        // validator_2 gets 40 SUI of their own rewards and 10 SUI of commission
-        // so in total 225 + 40 + 10 = 275 SUI.
-        // validator_3 and validator_4 just get their regular shares (60 SUI and 80 SUI).
-        assert_validator_self_stake_amounts(validator_addrs(), vector[132, 275, 390, 520], scenario);
+        // 1320 can be nicely partitioned to double the staking pools
+        advance_epoch_with_reward_amounts(0, 1320, scenario);
+        // V1: 440, V2: 660, V3: 660, V4: 880
+        assert_validator_total_stake_amounts(validator_addrs(), vector[440 * MIST_PER_SUI, 660 * MIST_PER_SUI, 660 * MIST_PER_SUI, 880 * MIST_PER_SUI], scenario);
+        
+        // Staker 1 rewards in the recent distribution is 0.9 x (220 / 2) = 99 SUI
+        // Validator 1 rewards in the recent distribution is 220 - 99 = 121 SUI
+        
+        // Old exchange rate for validator 2: 1.0.
+        // New exchange rate for validator 2: 660/300
+        // Staker 2 amounts for 0.5 * (660/300 * 100 - 100) + 100 = 160 SUI
+        // Validator 2 amounts for 660 - 160 = 500 SUI
+        // TODO: Investigate the discrepancy
```

Comment:
@emmazzz could you please help me with the investigation of this test? I'm not sure but there seems to be a problem with the calculations (my napkin calculations don't match the results)

Reply:
Hey @dmitri-perelman if we forget about the whole idea of exchange rate, and just think about how much staker 2 and validator 2 _should_ get, we see that the numbers do match:
- Validator 2 gets her share of the rewards plus half of staker 2's so that's 225 + 105/2 = 277.5 SUI in rewards. And together with the 225 SUI she already has, she gets 225 + 277.5 = 502.5 SUI, which is super close to 502499999999 MIST.
- Staker 2 therefore gets 105 * 1.5 = 157.5 SUI in total.

Now back to looking at the exchange rate, the issue is with the new exchange rate for validator 2. It's not 660/300. When commissions are distributed, they aren't added as rewards to the pool. Instead they are added as _new stakes_ of validators (this way the commission isn't shared with other stakers) to the pool. Also these commissions include commissions on the validator's self stake as well because, remember, validator is simply another staker. 

Therefore we did get 0.5 * 30 = 15 more SUI of stake added to the pool the first time and 0.5 * 330 = 165 SUI added the second time. The first time that translated to 15 * (300 / (300 + 15)) = 14.28 pool tokens and the second time that translated to 165 * (314.28 / (330 + 165)) = 104.76 pool tokens. So the number of pool tokens we have in the pool after the second reward distribution is 300 + 14.28 + 104.76 = 419.04. And plugging in the formula we get:
- Staker 2 amounts for 100 * (660 / 419.04) = 157.5 
- Validator 2 amounts for 660 - 157.5 = 502.5

Reply:
Wow. @emmazzz your math is strong